### ISSUE TITLE
difference between 3 training scenarios

### ISSUE THREAD

**Issue by MasIgor** [ultralytics/yolov5#8296]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello!

I'd like to present you 3 different training scenarios, and understand what the difference is between each of those. if it affects the training and how...

1) set up training and let it finish. For example 1000 epochs. 

2) set up training, abort it after 200 epochs, download the last.pt file from colab, upload it to a new colab the next day, start the training again with the previously downloaded last.pt file. it will continue from where it left the day before, but its missing all the other files present in the runs/train folder, such as best.pt, hyp.yaml, opt.yaml etc..

3) set up a training and let it finisch. For example 200 epocs. This wil finalize/optimize the last.pt file and reduce its size, stripping out a lot of data. Download the optimized last.pt file, and upload it the next day to a new colab. start a new training with the remianing 800 epochs. it will start from 0 this time. 

Is there a difference in the outcome of these 3 situations? what is the best/worst case?

also, is there a way to manually finalize a .pt file if the training is aborted prematurely?

Thank you!

Best regards

**Comment by mertz1999** [comment]
For number 2:
I recently do this. for example every 200 epoch you can save your model but you need to save DataLoader checkpoint, loss checkpoint and some other.

please see my work here:
https://colab.research.google.com/drive/1wzSPj403pbZ9Uwcc5E8G_bNqMqVxAlfG?usp=sharing
in Save model, Load prev. and Training Loop sections.

**Comment by glenn-jocher** [comment]
@tanzerlana 👋 Hello! Thanks for asking about **training checkpoints**.  YOLOv5 🚀 checkpoints should be about 4X the size of final trained checkpoints, as they carry not just a FP16 model, but a FP16 [EMA](https://www.tensorflow.org/api_docs/python/tf/train/ExponentialMovingAverage) and an FP32 optimizer of the same size as the model (each model parameter has its own FP32 gradient saved within the optimizer). Final checkpoints contain only an FP16 model, with the EMA and optimizer both stripped after the final epoch of training:
https://github.com/ultralytics/yolov5/blob/b4a29b5a8d63a8c2d4a8929942b44e8969c5dddd/train.py#L423-L425

The `strip_optimizer()` function updates the checkpoint dictionary value for `model`, replacing it with `ema`, and sets `ema` and `optimizer` keys to `None`, which will reduce the checkpoint size by 3/4.
https://github.com/ultralytics/yolov5/blob/b4a29b5a8d63a8c2d4a8929942b44e8969c5dddd/utils/general.py#L741-L755

Final trained checkpoints after `strip_optimizer()` should match the README table sizes below.
<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

You can also run `strip_optimizer()` manually on any checkpoint to convert it into finalized weights ready to train a new model:
```python
from utils.general import strip_optimizer

strip_optimizer('path/to/best.pt')
!python train.py --weights path/to/best.pt  # use best.pt to train a new model
```

### Model Exports

Exported models may be larger or smaller than their PyTorch starting points depending on their quantization. i.e. FP16 models will be the same size, FP32 models will be 2X the size, and INT8 models will be 0.5X the size. See YOLOv5 [Export tutorial](https://docs.ultralytics.com/yolov5/tutorials/model_export) for quantization options on model export.

Good luck 🍀 and let us know if you have any other questions!

**Comment by MasIgor** [comment]
thank you very much for yor detailed answer!

Am I understanding it correctly then, that trainings examples 1 and 2 in my question have identiacal results, and training 3 will have unpredictable results since its a new training with a different starting point compared of what the other two had?

**Comment by glenn-jocher** [comment]
@tanzerlana 👋 Hello! Thanks for asking about **resuming training**. YOLOv5 🚀 Learning Rate (LR) [schedulers](https://github.com/ultralytics/yolov5/blob/aa542ce6a65658ff931fee9bbab77c0145c152f0/train.py#L175-L181) follow predefined LR curves for the fixed number of `--epochs` defined at training start (`default=300`), and are designed to fall to a minimum LR on the final epoch for best training results. For this reason you can not modify the number of epochs once training has started.

<img width="850" alt="Screenshot 2022-04-10 at 11 11 51" src="https://user-images.githubusercontent.com/26833433/162611202-4e0b611f-6cca-4e43-94fc-bbc44e6447f8.png">

If your training was **interrupted** for any reason you may continue where you left off using the `--resume` argument. If your training fully completed, you can start a **new** training from any model using the `--weights` argument. Examples:

## Resume Single-GPU

You may not change settings when resuming, and **no additional arguments** other than `--resume` should be passed, with an optional path to the checkpoint you'd like to resume from. If no checkpoint is passed the most recently updated `last.pt` in your `yolov5/` directory is automatically found and used:
```bash
python train.py --resume  # automatically find latest checkpoint (searches yolov5/ directory)
python train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Resume Multi-GPU

[Multi-GPU DDP trainings](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training) must be resumed with the same GPUs and DDP command, i.e. assuming 8 GPUs:
```bash
python -m torch.distributed.run --nproc_per_node 8 train.py --resume  # resume latest checkpoint
python -m torch.distributed.run --nproc_per_node 8 train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Start from Pretrained

If you would like to **start training** from a fully trained model, use the `--weights` argument, not the `--resume` argument:
```bash
python train.py --weights path/to/best.pt  # start from pretrained model
```

Good luck 🍀 and let us know if you have any other questions!
